processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,151 @@
+ */
+
+@ManageLifecycle
+public class LookupReferencesManager
+{
+  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
+  private final ConcurrentMap<String, LookupExtractor> lookupMap = new ConcurrentHashMap();
+  private final Object lock = new Object();
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (lock) {
+      if (!started.getAndSet(true)) {
+        LOGGER.info("Started lookup references manager");
+      }
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (started.getAndSet(false)) {
+        for (String lookupName : lookupMap.keySet()) {
+          remove(lookupName);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param lookupName      namespace of the lookup object
+   * @param lookupExtractor {@link LookupExtractor} implementation reference.
+   *
+   * @return true if the lookup is added otherwise false.
+   *
+   * @throws ISE If the manager is closed.
+   */
+  public boolean addLookup(String lookupName, final LookupExtractor lookupExtractor) throws ISE
+  {
+    synchronized (lock) {
+      assertStarted();
+      return (null == lookupMap.putIfAbsent(lookupName, lookupExtractor));
+    }
+  }
+
+  /**
+   * @param lookupName namespace of {@link LookupExtractor} to delete from the reference registry.
+   *                   this function does call the cleaning method {@link LookupExtractor#close()}
+   *
+   * @return true if it is removed
+   */
+  public boolean remove(String lookupName)
+  {
+    final LookupExtractor lookupExtractor = lookupMap.remove(lookupName);
+    if (lookupExtractor != null) {
+      try {
+        LOGGER.debug("Removing lookup [%s]", lookupName);
+        lookupExtractor.close();
+      }
+      catch (IOException e) {
+        LOGGER.error(e, "Got exception while closing lookup [%s]", lookupName);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @param lookupName namespace key to fetch the reference of the object {@link LookupExtractor}
+   *
+   * @return reference of {@link LookupExtractor} that correspond the the namespace {@code lookupName}
+   *
+   * @throws ISE if the {@link LookupReferencesManager} is closed or did not start yet
+   */
+  public LookupExtractor get(String lookupName) throws ISE
+  {
+    synchronized (lock) {
+      assertStarted();
+      return lookupMap.get(lookupName);
+    }
+  }
+
+  /**
+   * @return Returns {@link ImmutableMap} containing a copy of the current state.
+   *
+   * @throws ISE if the is is closed or not started yet.
+   */
+  public ImmutableMap<String, LookupExtractor> getAll() throws ISE
+  {
+    synchronized (lock) {
+      assertStarted();
+      return ImmutableMap.copyOf(lookupMap);
+    }
+  }
+
+  private void assertStarted() throws ISE
+  {
+    if (isClosed()) {
+      throw new ISE("lookup manager is closed");
+    }
+  }
+
+  public boolean isClosed()
```

Comment:
i see, probably for unit tests , but then you don't need to make it public, default should be good enough. will be nice to use @VisibleForTesting annotation.